### ISSUE TITLE
External Script Backend

### ISSUE THREAD

**Issue by hongkongkiwi** [glauth/glauth#82]
Hi, for the backend implementation, can we simply have an external script called?

I can write a bash script to call a mongodb, external rest api, google storage etc. This way you don't need to do much work to handle the implementation of different backends, the user can write their own script to return the data in a specific format.

**Comment by benyanke** [comment]
I'd absolutely be open to an implementation of this.

**Comment by hongkongkiwi** [comment]
I'm keen but unfortunately I'm not a Go programmer (otherwise I would have added it :-) ) but I can certainly include lots of useful scripts once the feature is available.

**Comment by Fusion** [comment]
Reviving this: what do you think the best language for a scripted backend would be? Lua typically comes to mind, but maybe Javascript, or even a Go-like scripting language?

**Comment by hongkongkiwi** [comment]
How about calling a system executable OR calling a rest call so it can be backend agnostic?

**Comment by benyanke** [comment]
the powerdns "pipe" backend options might be a useful example here:

https://doc.powerdns.com/authoritative/backends/pipe.html

You can specify an arbitrary executable that is fed the incoming query, and responds back with the result. You can then use that script to implement whatever custom logic you want, whether lua, go, etc.

**Comment by Fusion** [comment]
Yes @benyanke I imagine these backends are communicating with the server through pipes, actually, which would mean not having to pay the fork/exec tax of executing an external binary.

@hongkongkiwi REST or gRPC, etc calls are also a good approach. Ironically when I was writing plugins for database support for GLAuth, I stumbled upon Hashicorp's version of plugins, which works similarly (the plugin runs in parallel and communicates with the main process through gRPC) - I ended up not adopting it for pure performance reasons, while, on the other hand, the list of benefits is quasi endless, especially when compared to Go plugins.
